Bugfix/fix events emitting and run handlers of active renderer only

milax opened the pull request:
Fixed another issue in Mediaelement with wrong events emitting. I found that all previously initialized renderers were handling events even when they are not active at the moment. It triggered random errors and the worst picture was, of course, in IE where error was thrown and player tries to reload the same resource. Thanks to Edge I noticed that when it was starting playing HLS (html5 renderer) and RTMP (flash_video renderer) at the same time, so user could listen to both livestream and regular track in parallel.

I didn't create an issue because I coudn't send a link to site where it's reproduced.

rafa8626 commented:
Thanks I'll review this right now. Is this a problem only with those renderers or it needs to be checked in the rest of the renderers, specially the native ones such as dash, hls and flv?

rafa8626 commented:
I see a change on  build/mediaelement-flash-video.swf but I don't see changes on `src/flash/flash-video/VideoMediaElement.as` is that expected?

rafa8626 commented:
Do you have a code snippet that I can use for testing purposes on the issue you had?

rafa8626 commented:
Nevermind I used the demo file for testing purposes. One thing I did notice is this: select the audio RTMP and play it. Then try to move ahead clicking the progress bar. The audio restarts and even with a Restart icon. Now play the video MP4 and you will see that both of them are playing simultaneously. So there are some events not being triggered correctly still

milax commented:
Hi @rafa8626 
Sorry, I will answer on your questions tomorrow, today I do not have time :(

rafa8626 commented:
Yeah no worries thanks

milax commented:
> Thanks I'll review this right now. Is this a problem only with those renderers or it needs to be checked in the rest of the renderers, specially the native ones such as dash, hls and flv?

I noticed the issue in Edge browser and tested only with these two renderers. In Chrome "native_hls" is used instead of "html5" for playing HLS but I didn't see the problem with it. So I didn't apply any fixes to other renderers because I have no chance to test them but I think the same fix might be useful as a safe guard to avoid triggering/handling playback events in a renderer which is not active at the moment.

milax commented:
> I see a change on build/mediaelement-flash-video.swf but I don't see changes on src/flash/flash-video/VideoMediaElement.as is that expected?
Actually I do not know why it's marked as changed. I've decompiled both versions from master and this bugfix branch and source code are absolutely similar. During work on the issue I changed VideoMediaElement.as couple of times and compiled to mediaelement-flash-video.swf but then I reverted those changes. And recompiled flash file again.

milax commented:
> Nevermind I used the demo file for testing purposes. One thing I did notice is this: select the audio RTMP and play it. Then try to move ahead clicking the progress bar. The audio stops, and displays a Restart icon. You click the restart icon and audio plays but if you click the progress bar again, it stops and starts from zero. Let it play and now play the video MP4 and you will see that both of them are playing simultaneously. So there are some events not being triggered correctly still. I don't know if the first issue is because of the RTMP source I have as an example

Yes, I see the problem. Investigating.

milax commented:
@rafa8626 
Can you reproduce it here http://htmltricks.ru/test/mediaelement/fixes/demo/ ? I can't :( The page contains a version from this branch.
For comparing I also added the latest version from master branch (it's not releases but contains some important fixes): http://htmltricks.ru/test/mediaelement/master/demo/

rafa8626 commented:
Thanks for investigating this. I'll check your demo as soon as I can and get back to you with my feedback

rafa8626 commented:
I have checked both of your pages, and you are not experiencing this because the audio RTMP has an error on your side but not on mine. I don't know if it's because of the region you are or what, but I tried them locally and I can reproduce the error I mentioned above. On your site, this is what I got:
<img width="1433" alt="screen shot 2017-08-27 at 10 11 14 pm" src="https://user-images.githubusercontent.com/910829/29756814-e6cb64aa-8b74-11e7-91e6-e938cb4ddbf9.png">

milax commented:
@rafa8626 hm, I haven't noticed such error both by the link (please watch the video https://youtu.be/wPqWc7AQewI) and on the site I'm working on locally.
Here is the latest master (Friday state)  - https://youtu.be/NjUlaO7QBoA
And video from demo on http://www.mediaelementjs.com/ - https://youtu.be/fpp4iuNxPSY

I use Chrome on Windows Version 60.0.3112.101 (Official Build) (64-bit)

rafa8626 commented:
I see. Well the error I saw didn't appear on your side, but in both videos you were able to see what I described: if you touch the progress bar you basically stop the audio and it resets. Any idea why this happens?

milax commented:
@rafa8626 after a brief look at the issue it looks like data from rendered in DOM UI player passed wrong to flash player and back. "seeked" and "seeking" events are triggered ok but they are not handled for some reason. I use my own UI for the player that's why I didn't notice this issue and it works fine at my project. The only thing I can say - that's another issue to investigate and it was not made in this PR, I've uploaded one more version to my host and you can ensure that there is absolutely the same behaviour within v4.2.4 http://htmltricks.ru/test/mediaelement/4.2.4/demo/

rafa8626 commented:
Thanks for the quick response. I'll merge your change, and hope you can provide a PR to solve this last issue in the near future.

rafa8626 commented:
PR merged. Thanks

rafa8626 commented:
@milax Just as a follow-up: were you planning to create a PR to solve the issue we mentioned above with the RTMP audio?

milax commented:
@rafa8626 I'll be back from vacation in two weeks, only then I'll be able to take a look at the issue

rafa8626 commented:
Sounds good thanks @milax
